File: AdSwitcher-iOS/AdSwitcher/Adapters/MaioAdapter/MaioAdapter.m
//
//  MaioAdapter.m
//  AdSwitcher
//
//  Created by tkyaji on 2016/07/04.
//  Copyright © 2016年 adwitcher. All rights reserved.
//

#import "MaioAdapter.h"

@implementation MaioAdapter {
    BOOL _isSkipped;
}

@synthesize interstitialAdDelegate;

#pragma - InterstitialAdAdapter

- (void)interstitialAdInitialize:(UIViewController *)viewController parameters:(NSDictionary<NSString *,NSString *> *)parameters testMode:(BOOL)testMode {

    NSString *mediaId = [parameters objectForKey:@"media_id"];
    
    _DLOG(@"mediaId=%@", mediaId);
    
    [Maio setAdTestMode:testMode];
    [Maio startWithMediaId:mediaId delegate:self];
}

- (void)interstitialAdLoad {
    if ([Maio canShow]) {
        _DLOG("ready");
        [self.interstitialAdDelegate interstitialAdLoaded:self result:YES];
        
    } else {
        [self adLoad:1];
    }
}

- (void)interstitialAdShow {
    _DLOG();
    [Maio show];
    _isSkipped = NO;
}



#pragma - MaioDelegate

- (void)maioWillStartAd:(NSString *)zoneId {
    _DLOG(@"zoneId:%@", zoneId);
    [self.interstitialAdDelegate interstitialAdShown:self];
}

- (void)maioDidFinishAd:(NSString *)zoneId playtime:(NSInteger)playtime skipped:(BOOL)skipped rewardParam:(NSString *)rewardParam {
    _DLOG(@"zoneId:%@ playtime:%zd skipped:%d rewardParam:%@", zoneId, playtime, skipped, rewardParam);
    _isSkipped = skipped;
}

- (void)maioDidClickAd:(NSString *)zoneId {
    _DLOG(@"zoneId:%@", zoneId);
    [self.interstitialAdDelegate interstitialAdClicked:self];
}

- (void)maioDidFail:(NSString *)zoneId reason:(MaioFailReason)reason {
    _DLOG(@"zoneId:%@ reason:%ld", zoneId, (long)reason);
    [self.interstitialAdDelegate interstitialAdClosed:self result:NO isSkipped:NO];
}

- (void)maioDidCloseAd:(NSString *)zoneId {
    _DLOG(@"zoneId:%@", zoneId);
    [self.interstitialAdDelegate interstitialAdClosed:self result:YES isSkipped:_isSkipped];
}


#pragma - Private methods

- (void)adLoad:(int)count {
    _DLOG("count=%d", count);
    
    dispatch_time_t time = dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC);
    dispatch_after(time, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        if ([Maio canShow]) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.interstitialAdDelegate interstitialAdLoaded:self result:YES];
            });
        } else if (count == 5) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.interstitialAdDelegate interstitialAdLoaded:self result:NO];
            });
        } else {
            [self adLoad:count + 1];
        }
    });
}

@end

File: AdSwitcher-Android/maio-adapter/src/main/java/net/adswitcher/adapter/maio/MaioAdapter.java
package net.adswitcher.adapter.maio;

import android.app.Activity;
import android.os.Handler;
import android.util.Log;

import java.util.Map;

import net.adswitcher.adapter.InterstitialAdAdapter;
import net.adswitcher.adapter.InterstitialAdListener;
import jp.maio.sdk.android.FailNotificationReason;
import jp.maio.sdk.android.MaioAds;
import jp.maio.sdk.android.MaioAdsListener;

/**
 * Created by tkyaji on 2016/07/20.
 */
public class MaioAdapter extends MaioAdsListener implements InterstitialAdAdapter {

    private static final String TAG = "MaioAdapter";

    private Activity activity;
    private InterstitialAdListener interstitialAdListener;
    private String zoneId;
    private boolean isSkipped;


    @Override
    public void interstitialAdInitialize(Activity activity, InterstitialAdListener interstitialAdListener, Map<String, String> parameters, boolean testMode) {
        this.interstitialAdListener = interstitialAdListener;
        this.activity = activity;

        String mediaId = parameters.get("media_id");
        this.zoneId = parameters.get("zone_id");

        Log.d(TAG, "videoAdInitialize : media_id=" + mediaId + ", zone_id=" + this.zoneId);

        MaioAds.setAdTestMode(testMode);
        MaioAds.init(activity, mediaId, this);
    }

    @Override
    public void interstitialAdLoad() {
        Log.d(TAG, "interstitialAdLoad");

        if (this.zoneId == null) {
            if (MaioAds.canShow(this.zoneId)) {
                this.interstitialAdListener.interstitialAdLoaded(MaioAdapter.this, true);

            } else {
                this.adLoad(1);
            }
        } else {
            if (MaioAds.canShow()) {
                this.interstitialAdListener.interstitialAdLoaded(MaioAdapter.this, true);

            } else {
                this.adLoad(1);
            }
        }
    }

    @Override
    public void interstitialAdShow() {
        this.isSkipped = false;

        if (this.zoneId == null) {
            MaioAds.show();
        } else {
            MaioAds.show(this.zoneId);
        }
    }

    private boolean canShow() {
        if (this.zoneId == null) {
            return MaioAds.canShow();
        } else {
            return MaioAds.canShow(this.zoneId);
        }
    }

    private void show() {
        if (this.zoneId == null) {
            MaioAds.show();
        } else {
            MaioAds.show(this.zoneId);
        }
    }


    @Override
    public void onInitialized() {
        Log.d(TAG, "onInitialized");
    }

    @Override
    public void onChangedCanShow(String zoneId, boolean newValue) {
        Log.d(TAG, "onChangedCanShow : zone_id=" + zoneId + ", new_value=" + newValue);
    }

    @Override
    public void onOpenAd(String zoneId) {
        Log.d(TAG, "onOpenAd : zone_id=" + zoneId);
    }

    @Override
    public void onStartedAd(String zoneId) {
        Log.d(TAG, "onStartedAd : zone_id=" + zoneId);
        this.interstitialAdListener.interstitialAdShown(this);
    }

    @Override
    public void onFinishedAd(int playtime, boolean skipped, int duration, String zoneId) {
        Log.d(TAG, "onFinishedAd : playtime=" + playtime + ", skipped=" + skipped + " duration=" + duration + ", zone_id=" + zoneId);
        this.isSkipped = skipped;
    }

    @Override
    public void onClosedAd(String zoneId) {
        Log.d(TAG, "onClosedAd : zone_id=" + zoneId);
        this.interstitialAdListener.interstitialAdClosed(this, true, this.isSkipped);
    }

    @Override
    public void onClickedAd(String zoneId){
        Log.d(TAG, "onClickedAd : zone_id=" + zoneId);
        this.interstitialAdListener.interstitialAdClicked(this);
    }

    @Override
    public void onFailed(FailNotificationReason reason, String zoneId) {
        Log.d(TAG, "onFailed : reason=" + reason + ", zone_id=" + zoneId);
    }


    private void adLoad(final int count) {
        Log.d(TAG, "adLoad : count=" + count);

        (new Handler()).postDelayed(new Runnable() {
            @Override
            public void run() {
                if (MaioAds.canShow()) {
                    MaioAdapter.this.activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            MaioAdapter.this.activity.runOnUiThread(new Runnable() {
                                @Override
                                public void run() {
                                    MaioAdapter.this.interstitialAdListener.interstitialAdLoaded(MaioAdapter.this, true);
                                }
                            });
                        }
                    });
                } else if (count == 5) {
                    MaioAdapter.this.activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            MaioAdapter.this.activity.runOnUiThread(new Runnable() {
                                @Override
                                public void run() {
                                    MaioAdapter.this.interstitialAdListener.interstitialAdLoaded(MaioAdapter.this, false);
                                }
                            });
                        }
                    });
                } else {
                    MaioAdapter.this.adLoad(count + 1);
                }
            }
        }, 1000);
    }

}
